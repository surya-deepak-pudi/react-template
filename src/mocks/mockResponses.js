export const fetch_users_mock = {
    data: {
        page: 1,
        per_page: 6,
        total: 12,
        total_pages: 2,
        data: [
            {
                id: 1,
                email: "george.bluth@reqres.in",
                first_name: "George",
                last_name: "Bluth",
                avatar: "https://reqres.in/img/faces/1-image.jpg",
            },
            {
                id: 2,
                email: "janet.weaver@reqres.in",
                first_name: "Janet",
                last_name: "Weaver",
                avatar: "https://reqres.in/img/faces/2-image.jpg",
            },
            {
                id: 3,
                email: "emma.wong@reqres.in",
                first_name: "Emma",
                last_name: "Wong",
                avatar: "https://reqres.in/img/faces/3-image.jpg",
            },
            {
                id: 4,
                email: "eve.holt@reqres.in",
                first_name: "Eve",
                last_name: "Holt",
                avatar: "https://reqres.in/img/faces/4-image.jpg",
            },
            {
                id: 5,
                email: "charles.morris@reqres.in",
                first_name: "Charles",
                last_name: "Morris",
                avatar: "https://reqres.in/img/faces/5-image.jpg",
            },
            {
                id: 6,
                email: "tracey.ramos@reqres.in",
                first_name: "Tracey",
                last_name: "Ramos",
                avatar: "https://reqres.in/img/faces/6-image.jpg",
            },
        ],
        support: {
            url: "https://reqres.in/#support-heading",
            text: "To keep ReqRes free, contributions towards server costs are appreciated!",
        },
    },
    status: 200,
    statusText: "OK",
    headers: {
        "access-control-allow-origin": "*",
        age: "1760",
        "cache-control": "max-age=14400",
        "cf-cache-status": "HIT",
        "cf-ray": "7fd65e72d9079aa5-NAG",
        connection: "close",
        "content-encoding": "br",
        "content-type": "application/json; charset=utf-8",
        date: "Sun, 27 Aug 2023 18:27:30 GMT",
        etag: 'W/"3e4-2RLXvr5wTg9YQ6aH95CkYoFNuO8"',
        nel: '{"success_fraction":0,"report_to":"cf-nel","max_age":604800}',
        "report-to":
            '{"endpoints":[{"url":"https:\\/\\/a.nel.cloudflare.com\\/report\\/v3?s=2abTqzfrg3SM9nWlEANOXBAAQRP6HKK%2F%2FJBwf57q19tBqhoPQDxvpMwbRbnqV%2F9uvl1Whj61mw%2B3OX71C5V51JND9nleT%2FUO7BDP1IhdBWB%2B6v2zcz%2BuI%2FKuzA%3D%3D"}],"group":"cf-nel","max_age":604800}',
        server: "cloudflare",
        "transfer-encoding": "chunked",
        vary: "Accept-Encoding",
        via: "1.1 vegur",
        "x-powered-by": "Express",
    },
    config: {
        transitional: {
            silentJSONParsing: true,
            forcedJSONParsing: true,
            clarifyTimeoutError: false,
        },
        adapter: "xhr",
        transformRequest: [null],
        transformResponse: [null],
        timeout: 0,
        xsrfCookieName: "XSRF-TOKEN",
        xsrfHeaderName: "X-XSRF-TOKEN",
        maxContentLength: -1,
        maxBodyLength: -1,
        env: {},
        headers: {
            Accept: "application/json, text/plain, */*",
        },
        method: "get",
        url: "api/users",
        signal: {},
    },
    request: {},
}
