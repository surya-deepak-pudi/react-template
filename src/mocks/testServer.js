import { setupServer } from "msw/node"
import { handlers } from "./handlers"

export const testServer = setupServer(...handlers)
