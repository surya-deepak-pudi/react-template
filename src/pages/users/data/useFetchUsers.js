import { useAPICall } from "Hooks"
import { defaultError } from "Utils/errors"
import { fetchUsersAPI } from "Utils/api"

const useFetchUsers = () => {
    const { data, setData, setError, error, isLoading, callAPI } = useAPICall()

    const defaultFallback = () => {
        setError(defaultError)
    }

    const statuses = [
        {
            status_id: 200,
            status_txt: "OK",
            callBack: (res) => {
                setData(res.data)
            },
        },
    ]

    const fetchUsers = () => {
        callAPI({
            url: fetchUsersAPI,
            defaultFallback,
            statuses,
        })
    }

    return {
        users: data,
        usersLoading: isLoading,
        usersError: error,
        fetchUsers,
    }
}

export default useFetchUsers
