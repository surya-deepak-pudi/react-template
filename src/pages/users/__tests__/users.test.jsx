//react imports
import React from "react"
//testing imports
import {
    fireEvent,
    getByText,
    render,
    screen,
    waitFor,
} from "@testing-library/react"
import renderer from "react-test-renderer"
import { rest } from "msw"
//testing helper imports
import { testServer } from "Mocks/testServer"
import { fetch_users_mock } from "Mocks/mockResponses"
//component file imports
import Users from "../index"
import * as useFetchUsers from "../data/useFetchUsers"

//sample users list
const users = [
    {
        id: 1,
        email: "george.bluth@reqres.in",
        first_name: "George",
        last_name: "Bluth",
        avatar: "https://reqres.in/img/faces/1-image.jpg",
    },
    {
        id: 2,
        email: "janet.weaver@reqres.in",
        first_name: "Janet",
        last_name: "Weaver",
        avatar: "https://reqres.in/img/faces/2-image.jpg",
    },
    {
        id: 3,
        email: "emma.wong@reqres.in",
        first_name: "Emma",
        last_name: "Wong",
        avatar: "https://reqres.in/img/faces/3-image.jpg",
    },
    {
        id: 4,
        email: "eve.holt@reqres.in",
        first_name: "Eve",
        last_name: "Holt",
        avatar: "https://reqres.in/img/faces/4-image.jpg",
    },
    {
        id: 5,
        email: "charles.morris@reqres.in",
        first_name: "Charles",
        last_name: "Morris",
        avatar: "https://reqres.in/img/faces/5-image.jpg",
    },
    {
        id: 6,
        email: "tracey.ramos@reqres.in",
        first_name: "Tracey",
        last_name: "Ramos",
        avatar: "https://reqres.in/img/faces/6-image.jpg",
    },
]

//test suite for users page
describe("Testing the Users Screen", () => {
    // fetch test example
    it("Should render the title of the page", () => {
        render(<Users></Users>)
        const title = screen.queryByTestId("user-title")
        expect(title).toBeInTheDocument()
        expect(getByText(title, "Users")).toBeTruthy()
    })
    //fire-event example
    it("Should render the fetch button", () => {
        render(<Users></Users>)
        const button = screen.queryByTestId("user-fetch-button")
        expect(button).toBeInTheDocument()
        fireEvent.click(button)
    })
    //spy on example if ur using external function mocking
    it("Should show loading screen", async () => {
        jest.spyOn(useFetchUsers, "default").mockReturnValue({
            usersLoading: true,
            fetchUser: jest.fn(() => {}),
        })
        render(<Users></Users>)
        await waitFor(() => {
            const loader = screen.queryByTestId("user-loader-container")
            expect(loader).toBeInTheDocument()
        })
    })
    it("Should show error screen", async () => {
        jest.spyOn(useFetchUsers, "default").mockReturnValue({
            usersError: true,
            fetchUser: jest.fn(() => {}),
        })
        render(<Users></Users>)
        await waitFor(() => {
            const error = screen.queryByTestId("user-error")
            expect(error).toBeInTheDocument()
        })
    })
    it("Should show users screen", async () => {
        jest.spyOn(useFetchUsers, "default").mockReturnValue({
            users,
            fetchUser: jest.fn(() => {}),
        })
        render(<Users></Users>)
        await waitFor(() => {
            const list = screen.queryByTestId("user-list")
            expect(list).toBeInTheDocument()
        })
    })
    //test server example when a call happens internally
    it("Should show users screen with test server", async () => {
        testServer.use(
            rest.get("http://localhost/api/users", (req, res, ctx) => {
                return res(ctx.json(fetch_users_mock))
            })
        )
        render(<Users></Users>)
        await waitFor(() => {
            const list = screen.queryByTestId("user-list")
            expect(list).toBeInTheDocument()
        })
    })
    //snapshot testing example
    it("Should render users list with snapshot", () => {
        jest.spyOn(useFetchUsers, "default").mockReturnValue({
            users,
            fetchUser: jest.fn(() => {}),
        })
        const tree = renderer.create(<Users></Users>).toJSON()
        expect(tree).toMatchSnapshot()
    })
})
