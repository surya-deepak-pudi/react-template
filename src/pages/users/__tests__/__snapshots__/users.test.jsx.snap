// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Users Screen Should render users list with snapshot 1`] = `
<div
  className="MuiBox-root css-0"
>
  <p
    className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    data-testid="user-title"
  >
    Users
  </p>
  <div
    className="MuiBox-root css-0"
    data-testid="user-list"
  >
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-alignItemsFlexStart css-hhvo2v-MuiListItem-root"
      data-testid="user-list-item"
      disabled={false}
    >
      <div
        className="MuiListItemAvatar-root MuiListItemAvatar-alignItemsFlexStart css-1m6607n-MuiListItemAvatar-root"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
          data-testid="user-list-avatar"
        >
          <img
            alt="George"
            className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="https://reqres.in/img/faces/1-image.jpg"
          />
        </div>
      </div>
      <div
        className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        data-testid="user-list-text"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          George Bluth
        </span>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body2 css-1phucj-MuiTypography-root"
          >
            george.bluth@reqres.in
          </span>
        </p>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-alignItemsFlexStart css-hhvo2v-MuiListItem-root"
      data-testid="user-list-item"
      disabled={false}
    >
      <div
        className="MuiListItemAvatar-root MuiListItemAvatar-alignItemsFlexStart css-1m6607n-MuiListItemAvatar-root"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
          data-testid="user-list-avatar"
        >
          <img
            alt="Janet"
            className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="https://reqres.in/img/faces/2-image.jpg"
          />
        </div>
      </div>
      <div
        className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        data-testid="user-list-text"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Janet Weaver
        </span>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body2 css-1phucj-MuiTypography-root"
          >
            janet.weaver@reqres.in
          </span>
        </p>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-alignItemsFlexStart css-hhvo2v-MuiListItem-root"
      data-testid="user-list-item"
      disabled={false}
    >
      <div
        className="MuiListItemAvatar-root MuiListItemAvatar-alignItemsFlexStart css-1m6607n-MuiListItemAvatar-root"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
          data-testid="user-list-avatar"
        >
          <img
            alt="Emma"
            className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="https://reqres.in/img/faces/3-image.jpg"
          />
        </div>
      </div>
      <div
        className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        data-testid="user-list-text"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Emma Wong
        </span>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body2 css-1phucj-MuiTypography-root"
          >
            emma.wong@reqres.in
          </span>
        </p>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-alignItemsFlexStart css-hhvo2v-MuiListItem-root"
      data-testid="user-list-item"
      disabled={false}
    >
      <div
        className="MuiListItemAvatar-root MuiListItemAvatar-alignItemsFlexStart css-1m6607n-MuiListItemAvatar-root"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
          data-testid="user-list-avatar"
        >
          <img
            alt="Eve"
            className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="https://reqres.in/img/faces/4-image.jpg"
          />
        </div>
      </div>
      <div
        className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        data-testid="user-list-text"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Eve Holt
        </span>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body2 css-1phucj-MuiTypography-root"
          >
            eve.holt@reqres.in
          </span>
        </p>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-alignItemsFlexStart css-hhvo2v-MuiListItem-root"
      data-testid="user-list-item"
      disabled={false}
    >
      <div
        className="MuiListItemAvatar-root MuiListItemAvatar-alignItemsFlexStart css-1m6607n-MuiListItemAvatar-root"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
          data-testid="user-list-avatar"
        >
          <img
            alt="Charles"
            className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="https://reqres.in/img/faces/5-image.jpg"
          />
        </div>
      </div>
      <div
        className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        data-testid="user-list-text"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Charles Morris
        </span>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body2 css-1phucj-MuiTypography-root"
          >
            charles.morris@reqres.in
          </span>
        </p>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-alignItemsFlexStart css-hhvo2v-MuiListItem-root"
      data-testid="user-list-item"
      disabled={false}
    >
      <div
        className="MuiListItemAvatar-root MuiListItemAvatar-alignItemsFlexStart css-1m6607n-MuiListItemAvatar-root"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
          data-testid="user-list-avatar"
        >
          <img
            alt="Tracey"
            className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="https://reqres.in/img/faces/6-image.jpg"
          />
        </div>
      </div>
      <div
        className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        data-testid="user-list-text"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Tracey Ramos
        </span>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body2 css-1phucj-MuiTypography-root"
          >
            tracey.ramos@reqres.in
          </span>
        </p>
      </div>
    </li>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
    data-testid="user-fetch-button"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Fetch!
  </button>
</div>
`;
